datafusion/expr-common/src/scalar.rs

```diff
@@ -0,0 +1,109 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::{
+    array::{Array, ArrayRef},
+    datatypes::DataType,
+};
+use datafusion_common::{exec_err, DataFusionError, Result, ScalarValue};
+
+#[derive(Clone, Debug)]
+pub struct Scalar {
+    value: ScalarValue,
+    data_type: DataType,
```

Comment:
Should I assume that constructing `Scalar { value:  ScalarValue::Int32(..), data_type: DataType::Null }` is not valid?
We need to define constraints on these values and ideally enforce them.

Reply:
I think those constraint can be made implicit if we consider editing the possible way the `Scalar` object can be constructed. By only keeping the `::from(ScalarValue)` and the `::try_from_array(...)` method (which is used for casting) we artificially limit the possible values that `data_type` can hold to the ones that are cast-compatible with `value`.

Reply:
Another potential way to represent this idea would be to use an enum, perhaps, like

```rust
pub enum Scalar {
  /// A ScalarValue (transitory)
  ScalarValue(ScalarValue),
  /// A logical String with associated Arrow physical typ
  String {
    value: Option<String>,
    data_type: DataType,
  },
}
```

I think we could modify this in the future as well

Reply:
@alamb  what would be "a logical String with associated data type"? i.e. what value of target data type does it represent and how to obtain it? And it it costly to obtain it during optimization process?
i didn't think about this too deeply yet, but would it be possible for the logical IR to avoid ambiguities?

> I think those constraint can be made implicit if we consider editing the possible way the `Scalar` object can be constructed. By only keeping the `::from(ScalarValue)` and the `::try_from_array(...)` method (

@notfilippo that's great there are constraints. 
let's make sure they are documented -- would be useful for the readers, and also would capture the design intent in the code.